### DISCUSSION TITLE
[Suggestions] Difference between the position of the elements in fheroes2 and the original game

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#2448] [Q&A]
In fheroes2, there are a lot of differences in the position of the elements compared to the original game, and I haven't checked all of that yet.

Wouldn't it be simpler, and ultimately faster, to look directly at the position of all the elements of the game in the source code and compare this in the fheroes2 code?

Because as it is, there might be dozens and dozens of new issues related to this, while doing it all at once would be much more convenient.

**Comment by ihhub** [comment]
The original game stores location of elements in bin files. The developers before us hardcoded values into source code without putting the correct data.
In this case correcting harcoded values or switching to use bin files (the format is still unknown fully) is not trivial.
Using separate files is not always more flexible and makes life of developers easier but we need time to check how to do it properly.

**Comment by Branikolog** [comment]
> The developers before us hardcoded values into source code without putting the correct data.

That's a bad news to hear, but I'm not surprised...
Do you want to say, that we have to rework all these elements in future?
Even if it's placed properly, but hardcoded?

**Reply by ihhub** [reply]
Yes.

**Comment by LeHerosInconnu** [comment]
Wouldn't it be possible to define the clickable areas automatically in the code based on the non-transparent pixels of the graphics?

**Reply by ihhub** [reply]
Everything is possible :)
Could you please give an example? Also there are 'transparent' pixels and 'transform' pixels like shadow.

**Reply by LeHerosInconnu** [reply]
Hello @ihhub,

> Everything is possible :)
> Could you please give an example? Also there are 'transparent' pixels and 'transform' pixels like shadow.

For example, for the Black Tower sprite, it can be considered that all the cyan colored pixels (transparencies) around it are not taken into account in the Black Tower clickable area and that all pixels other than the cyan colored ones define the clickable area, and the cyan colored pixels inside the Black Tower pixels (the "window") are added to the clickable area.

![Dragon tower 01](https://user-images.githubusercontent.com/43583994/109343278-610de100-786d-11eb-9a59-c9995ebf4203.png)

For magenta colored pixels (corresponding to the shadow), as for cyan colored pixels, they are not taken into account if they are located outside the displayed sprite area.

However, in some cases (here the Mage Guild with a road and a "shadow"), this would not be optimal and it would be necessary to define the clickable area (red rectangle in the example) in another way.

![Mage Guild with road avec cadre rouge 01](https://user-images.githubusercontent.com/43583994/109343291-679c5880-786d-11eb-8ceb-a0308465a024.png)

Or tell the program to take only part of the sprite to determine the clickable area.

![Mage Guild sprite detail 01](https://user-images.githubusercontent.com/43583994/109343304-6c610c80-786d-11eb-842a-2014811e51c2.png)

I don't know if this can be done automatically.

**Reply by ihhub** [reply]
As you mentioned this is not a trivial task and writing an algorithm to cover all cases might take so much time. Also in case of Dragon Tower the window inside the tower should be clickable as well.
The best and fastest way to identify building areas (rectangles) is to manually set them and then program into the game. This is what we have right now but the values are incorrect. We even have an issue for this case. We can definitely improve such cases by changing these values in the code.
If you have a spare time to identify such areas within separate images I can easily change the code. What we need is a position of top left corner of the rectangle (x, y) and the size of the rectangle (width, height) keeping in mind that first pixel coordinate is (0, 0). What do you think?

**Reply by LeHerosInconnu** [reply]
> As you mentioned this is not a trivial task and writing an algorithm to cover all cases might take so much time. Also in case of Dragon Tower the window inside the tower should be clickable as well.
> The best and fastest way to identify building areas (rectangles) is to manually set them and then program into the game. This is what we have right now but the values are incorrect. We even have an issue for this case. We can definitely improve such cases by changing these values in the code.
> If you have a spare time to identify such areas within separate images I can easily change the code. What we need is a position of top left corner of the rectangle (x, y) and the size of the rectangle (width, height) keeping in mind that first pixel coordinate is (0, 0). What do you think?

The goal is precisely not to have to do it "manually", because I don't really have the time to do it (and I don't really have the desire to do it either). :)

**Reply by ihhub** [reply]
Haha.
That's fine. I'm not forcing to do this. I think we can come up with some semi-hardcoded solution in the future.
